### PR TITLE
Fix & improve string error when quote is followed by eof

### PR DESCRIPTION
**Pull request by chenglou**
- Previously, `let a = "` followed by an eof would throw, despite `scanString` promising the opposite.
- This also includes the string's beginning quote into the error display (as opposed to only the content), so that such error doesn't highlight nothing.

### CODE REVIEW

File: src/res_scanner.ml [hunk 1]
```diff
@@ -350,26 +350,32 @@ let scanStringEscapeSequence ~startPos scanner =
     ()
 
 let scanString scanner =
-  let offs = scanner.offset in
+  (* assumption: we've just matched a quote *)
```
Comment (chenglou) [inline src/res_scanner.ml:353]: More dangerous here but it's ok. Only used in this file in one place.

File: src/res_scanner.ml [hunk 2]
```diff
@@ -350,26 +350,32 @@ let scanStringEscapeSequence ~startPos scanner =
...
+  let firstCharOffset = scanner.offset in
 
-  let startPos = position scanner in
   let rec scan () =
     match scanner.ch with
-    | '"' -> next scanner
+    | '"' ->
+      let lastCharOffset = scanner.offset in
+      next scanner;
+      (String.sub [@doesNotRaise]) scanner.src firstCharOffset (lastCharOffset - firstCharOffset)
```
Comment (chenglou) [inline src/res_scanner.ml:364]: No more `- 1` logic we need to double check!

File: tests/parsing/errors/scanner/__snapshots__/parse.spec.js.snap [hunk 3]
```diff
@@ -165,11 +165,11 @@ exports[`unclosedComment.js 1`] = `
 
 exports[`unclosedString.js 1`] = `
 "=====Parsetree==========================================
-let z = \\"eof\\"
+let z = \\"eof\\\\n\\"
```
Comment (chenglou) [inline tests/parsing/errors/scanner/__snapshots__/parse.spec.js.snap:168]: This is a natural change but one can argue both ways; whether to include the newline into the highlight or not

File: tests/parsing/errors/scanner/__snapshots__/parse.spec.js.snap [hunk 4]
```diff
@@ -165,11 +165,11 @@ exports[`unclosedComment.js 1`] = `
...
 =====Errors=============================================
 
   Syntax error!
-  parsing/errors/scanner/unclosedString.js:1:10-2:0  
+  parsing/errors/scanner/unclosedString.js:1:9-2:0
```
Comment (chenglou) [inline tests/parsing/errors/scanner/__snapshots__/parse.spec.js.snap:172]: @IwanKaramazow is there a case where we might need to highlight an ending `"`? I've only included the beginning `"`

File: src/res_scanner.ml (hunk 1 above)
Comment (IwanKaramazow) [reply src/res_scanner.ml:353]: Yes, this is an unofficial implicit assumption I regularly try to follow.

**Comment by chenglou** [comment]
Fixes part of #303

File: tests/parsing/errors/scanner/__snapshots__/parse.spec.js.snap (hunk 4 above)
Comment (IwanKaramazow) [reply tests/parsing/errors/scanner/__snapshots__/parse.spec.js.snap:172]: Hmm good question, my first reaction would be: unclosed strings will never have one. So we can't highlight it either way?

File: tests/parsing/errors/scanner/__snapshots__/parse.spec.js.snap (hunk 3 above)
Comment (IwanKaramazow) [reply tests/parsing/errors/scanner/__snapshots__/parse.spec.js.snap:168]: How does this look in the terminal?

Comment (chenglou) [reply tests/parsing/errors/scanner/__snapshots__/parse.spec.js.snap:168]: Before:
Screen Shot 2021-03-16 at 12 22 00 AM

After:
Screen Shot 2021-03-16 at 12 22 38 AM

No visible difference visually for the existing highlighter

Comment (IwanKaramazow) [reply tests/parsing/errors/scanner/__snapshots__/parse.spec.js.snap:168]: The newline might be an artefact from my editor. Everything before the end-of-file token should be a part.

**Pull request merged**